File: javascript/section01/solution08.js
/*
일곱난쟁이

왕비를 피해 일곱 난쟁이들과 함께 평화롭게 생활하던 백설공주
일과를 마치고 온 난쟁이가 일곱명이 아니라 아홉명이었던 것
아홉명의 난쟁이는 모두 자신이 " 백설공주와 일곱난쟁이"의 주인공이라 주장
일곱난쟁이들의 키의 합이 100이 됨을 기억한 백설공주
아홉난쟁이들이 주어졌을 때, 백설공주를 도와 일곱난쟁이를 찾는 프로그램

input : 20 7 23 19 10 15 25 8 13
output : 20 7 23 19 10 8 13
*/

let nine = [20, 7, 23, 19, 10, 15, 25, 8, 13];
console.log(solution(nine));

function solution(arr) {
    // 얕은 복사 : 메모리 주소를 복사 -> 똑같은 배열을 가리킴
    let answer = arr;
    // reduce() : a에 b를 누적 계산 (a는 0으로 초기화)
    let sum = arr.reduce((a, b) => a+b, 0);
    
    for(let i = 0; i < 8; i++) {
        for(let j = i+1; j < 9; j++) {
            if((sum - (arr[i] + arr[j])) === 100) {
                arr.splice(i, 1);
                arr.splice(j-1, 1);
            }
        }
    }

    return answer;
}

File: java/java_study/src/section02/Solution08.java
package section02;

import java.util.ArrayList;

public class Solution08 {

    public static void main(String[] args) {
        /*
        * 등수 구하기
        *
        * N명의 학생의 국어점수가 입력되면 각 학생의 둥수를 입력된 순서대로 출력하는 프로그램
        * 같은 점수가 입력될 경우 높은 등수로 동일 처리. 즉 가장 높은 점수가 92점인데 92점이 3명 존재하면 1등이 3명이고 그 다음 학생은 4등이 된다.
        *
        * input : 87, 89, 92, 100, 76
        * output : 4 3 2 1 5
        *
        * */

        int[] score = {87, 89, 92, 100, 76};
        System.out.print("result : ");
        for (int s : solution(score)) {
            System.out.print(s + " ");
        }
        System.out.println();
    }

    public static ArrayList<Integer> solution(int[] score) {
        ArrayList<Integer> answer = new ArrayList<>();

        for (int i = 0; i < score.length; i++) {
            // 1등부터 시작
            int cnt = 1;
            for (int j = 0; j < score.length; j++) {
                if (score[j] > score[i]) {
                    cnt++;
                }
            }
            answer.add(cnt);
        }

        return answer;
    }
}
